File: week6/ALL_PAIR_SHORTEST_PATH.cpp
//C++ 
#include <bits/stdc++.h>
using namespace std;

const int MAX_VERTICES = 111111;
const int MAX_EDGES = 311111;

vector<pair<int, int>> graph[MAX_EDGES];
bool visited[MAX_VERTICES];
int minDistance[MAX_VERTICES];
int numVertices, numEdges;

void applyDijkstra(int startVertex) {
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;
    fill(minDistance, minDistance + numVertices + 1, INT_MAX);
    fill(visited, visited + numVertices + 1, false);
    minDistance[startVertex] = 0;
    minHeap.push({0, startVertex});

    while (!minHeap.empty()) {
        int currentVertex = minHeap.top().second;
        minHeap.pop();

        if (!visited[currentVertex]) {
            for (auto neighbor : graph[currentVertex]) {
                int nextVertex = neighbor.first;
                int edgeWeight = neighbor.second;

                if (minDistance[nextVertex] > minDistance[currentVertex] + edgeWeight) {
                    minDistance[nextVertex] = minDistance[currentVertex] + edgeWeight;
                    minHeap.push({minDistance[nextVertex], nextVertex});
                }
            }
        }

        visited[currentVertex] = true;
    }
}

int main() {
    cin >> numVertices >> numEdges;

    for (int i = 0; i < numEdges; i++) {
        int source, destination, weight;
        cin >> source >> destination >> weight;
        graph[source].push_back({destination, weight});
    }

    for (int i = 1; i <= numVertices; i++) {
        applyDijkstra(i);

        for (int j = 1; j <= numVertices; j++) {
            if (i == j) {
                cout << 0 << " ";
            } else {
                cout << minDistance[j] << " ";
            }
        }

        cout << "\n";
    }

    return 0;
}

File: week6/ALL_PAIR_SHORTEST_PATH.java
//JAVA 
import java.util.*;

class Main {
    static final int MAX_VERTICES = 111111;
    static final int MAX_EDGES = 311111;

    static List<Edge>[] adjacencyList = new ArrayList[MAX_EDGES];
    static boolean[] visited = new boolean[MAX_VERTICES];
    static int[] minDistances = new int[MAX_VERTICES];
    static int numVertices, numEdges;

    static class Edge {
        int destination;
        int weight;

        Edge(int destination, int weight) {
            this.destination = destination;
            this.weight = weight;
        }
    }

    static void applyDijkstra(int startVertex) {
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        Arrays.fill(minDistances, Integer.MAX_VALUE);
        Arrays.fill(visited, false);
        minDistances[startVertex] = 0;
        minHeap.add(new int[]{0, startVertex});

        while (!minHeap.isEmpty()) {
            int[] current = minHeap.poll();
            int currentVertex = current[1];

            if (!visited[currentVertex]) {
                for (Edge neighbor : adjacencyList[currentVertex]) {
                    int nextVertex = neighbor.destination;
                    int edgeWeight = neighbor.weight;

                    if (minDistances[nextVertex] > minDistances[currentVertex] + edgeWeight) {
                        minDistances[nextVertex] = minDistances[currentVertex] + edgeWeight;
                        minHeap.add(new int[]{minDistances[nextVertex], nextVertex});
                    }
                }
            }

            visited[currentVertex] = true;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        numVertices = scanner.nextInt();
        numEdges = scanner.nextInt();

        for (int i = 1; i <= numVertices; i++) {
            adjacencyList[i] = new ArrayList<>();
        }

        for (int i = 0; i < numEdges; i++) {
            int source = scanner.nextInt();
            int destination = scanner.nextInt();
            int weight = scanner.nextInt();
            adjacencyList[source].add(new Edge(destination, weight));
        }

        for (int i = 1; i <= numVertices; i++) {
            applyDijkstra(i);

            for (int j = 1; j <= numVertices; j++) {
                if (i == j) {
                    System.out.print(0 + " ");
                } else {
                    System.out.print(minDistances[j] + " ");
                }
            }

            System.out.println();
        }
    }
}
